sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConsumer.cs

```diff
@@ -237,5 +348,270 @@ protected EventHubConsumer()
+        /// <param name="maximumQueuedEvents">The maximum number of events to queue while waiting for them to be read.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to signal the request to cancel the event enumeration.</param>
+        ///
+        /// <returns>The elements of a channel subscription, including its identifier and a reader for the subscribed channel.</returns>
+        ///
+        private (Guid Identifier, ChannelReader<EventData> ChannelReader) SubscribeToChannel(string eventHubName,
+                                                                                             string partitionId,
+                                                                                             string consumerGroup,
+                                                                                             int maximumQueuedEvents,
+                                                                                             CancellationToken cancellationToken)
+        {
+            var channel = CreateEventChannel(maximumQueuedEvents);
+            var identifier = Guid.NewGuid();
+
+            if (!ActiveChannels.TryAdd(identifier, channel))
+            {
+                throw new EventHubsException(true, eventHubName, String.Format(CultureInfo.CurrentCulture, Resources.FailedToCreateEventSubscription, eventHubName, partitionId, consumerGroup));
+            }
+
+            // Ensure that publishing is taking place.  To avoid race conditions, always acquire the semaphore and then perform the check.
+
+            ChannelSyncRoot.Wait(cancellationToken);
+
+            try
+            {
+                if (!_isPublishingActive)
+                {
+                    _channelPublishingTokenSource?.Cancel();
+                    _channelPublishingTask?.GetAwaiter().GetResult();
+                    _channelPublishingTokenSource?.Dispose();
+
+                    _channelPublishingTokenSource = new CancellationTokenSource();
+                    _channelPublishingTask = StartBackgroundChannelPublishingAsync(_channelPublishingTokenSource.Token);
+                    _isPublishingActive = true;
+                }
+            }
+            finally
+            {
+                ChannelSyncRoot.Release();
+            }
+
+            return (identifier, channel.Reader);
+        }
+
+        /// <summary>
+        ///   Unsubscribe from a channel subscription.
+        /// </summary>
+        ///
+        /// <param name="identifier">The identifier of the subscription to unsubscribe.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task UnsubscribeFromChannelAsync(Guid identifier)
+        {
+            if ((ActiveChannels.TryRemove(identifier, out var unsubscribeChannel)) && (ActiveChannels.Count == 0))
```

Comment:
What if we let the reading loop do this cleanup as soon as it detects 0 writers? Might require less syncronization.

Reply:
Not sure that I'm following your thought here.  That said,  I'm going to track this informally and defer exploring to not delay code-complete.  I'll sync up with you at some point next week and pick your brain.